Story/ver 92279 rebase save parquet files

jslaunwhite-microfocus opened the pull request:
## Query Profile Export

This PR:
* Adds import/export functionality to the QueryProfiler Object
* Adds tests of importing and exporting query profiles

It doesn't reproduce all aspects of query profile creation and collection. In particular, it is missing some tracking tables that can help to validate the rows collected, exported, and imported.

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -34,6 +36,8 @@
 from verticapy._utils._sql._vertica_version import vertica_version
 
 from verticapy.performance.vertica.tree import PerformanceTree
+from verticapy.performance.vertica.collection.profile_export import ProfileExport
```
A-Z order

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,75 @@ def get_cluster_config(self) -> vDataFrame:
+        -----------------
+        filename: os.PathLike
+           The name of the output file to produce. The output file
+           will be a tarball containing parquet files.
```
you need to build up an example

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,75 @@ def get_cluster_config(self) -> vDataFrame:
+
+    def export_profile(self, filename: os.PathLike) -> None:
+        """
+        Creates a tar file containing parquet files of the replica tables used while
```
shorter lines

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,75 @@ def get_cluster_config(self) -> vDataFrame:
+
+        """
+
+        if self.target_schema is None or len(self.target_schema) == 0:
```
it's better to import NoneType from verticapy typing and to use isinstance

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,75 @@ def get_cluster_config(self) -> vDataFrame:
+
+        if self.target_schema is None or len(self.target_schema) == 0:
+            raise ValueError(
+                f"Export requires that target_schema is set."
```
f is not require here

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,75 @@ def get_cluster_config(self) -> vDataFrame:
+        tmp_dir: os.PathLike = os.getenv("TMPDIR", "/tmp"),
+    ):
+        """
+        Import the query profile in filename into the database schema ``target_schema`` with tables
```
same than above

oualib commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,137 @@
+"""
```
@vikash018 please look at this file and tell us if it works for you

vikash018 replied:
@jslaunwhite-microfocus - As we discussed in call, we need to add following UT
1) Row count validation for import/export rows
2) Data content validation
3) Interval data validation for imported rows

Pls feel free to update above list for missing tests.

jslaunwhite-microfocus replied:
I added a simple row count validation for import/export. A previous PR already added some tests for import, but things changed enough that it is worth checking again.

oualib reviewed (changes requested):
Please review the changes.
It is very important to add an entire example for your functions.

jslaunwhite-microfocus reviewed (commented):
I have completed all the of the changes requested so far.

jslaunwhite-microfocus commented:
Builds are running. I've addressed all of the comments, and done some additional cleanup. Please take a look! I'd love to merge on Wednesday so that I can follow up with any issues on Thursday.

jslaunwhite-microfocus reviewed (commented):
All changes complete

jslaunwhite-microfocus commented:
@oualib  - can you please take a look and let me know if this is ok to move forward? I think it is ready to go ahead.

oualib commented on verticapy/performance/vertica/collection/collection_tables.py:
```diff
@@ -54,7 +57,61 @@ class BundleVersion(Enum):
+            "tables": [x.to_json() for x in self.tables],
+        }
+
+    def write_to_file(self):
```
We need precise description for each function. Many docstrings are missing in the PR.

jslaunwhite-microfocus replied:
Added a docstring for this function and class

oualib commented on verticapy/performance/vertica/collection/collection_tables.py:
```diff
@@ -54,7 +57,61 @@ class BundleVersion(Enum):
     """
 
     V1 = 1
-    LATEST = V1
```
what's that?

jslaunwhite-microfocus replied:
Updated class docstring and added comments to each version

oualib commented on verticapy/performance/vertica/collection/collection_tables.py:
```diff
@@ -1222,6 +1331,25 @@ def get_pandas_column_type_adjustments(self) -> Mapping[str, str]:
         }
 
+    def get_export_sql(self):
+        return f"""SELECT
```
Make this string more readable please

jslaunwhite-microfocus replied:
I ran out of time to change this formatting.

oualib commented on verticapy/performance/vertica/collection/collection_tables.py:
```diff
@@ -1489,3 +1617,46 @@ def get_pandas_column_type_adjustments(self) -> Mapping[str, str]:
         }
+
+    def get_export_sql(self):
+        return f"""SELECT
```
same

jslaunwhite-microfocus replied:
I didn't change the formating

oualib commented on verticapy/performance/vertica/collection/profile_export.py:
```diff
@@ -0,0 +1,163 @@
+)
+
+
+class ProfileExportError(Exception):
```
should it be described and added to the errors file?

jslaunwhite-microfocus replied:
I added a docstring to describe this error. I don't know how to add it to the errors file.

oualib commented on verticapy/performance/vertica/collection/profile_export.py:
```diff
@@ -0,0 +1,163 @@
+        filename: Path,
+    ) -> None:
+        """
+        Store common parameters for profile export
```
more complete docstring?

jslaunwhite-microfocus replied:
Updated this doc string

oualib commented on verticapy/performance/vertica/collection/profile_export.py:
```diff
@@ -0,0 +1,163 @@
+        export_metadata = self._save_tables()
+        self._bundle_tables(export_metadata)
+
+    def _tables_exist_or_raise(self) -> None:
```
many docstrings are missing...

jslaunwhite-microfocus replied:
I added a docstring for this function

oualib commented on verticapy/performance/vertica/collection/profile_export.py:
```diff
@@ -0,0 +1,163 @@
+        return existing_tables
+
+    def _save_tables(self) -> ExportMetadata:
+        """ """
```
same

jslaunwhite-microfocus replied:
I added a docstring for this function

oualib commented on verticapy/performance/vertica/qprof.py:
```diff
@@ -3283,3 +3287,116 @@ def get_cluster_config(self) -> vDataFrame:
+            from verticapy.performance.vertica import QueryProfiler
+            # The target_schema argument tells the QueryProfiler
+            # to create a set of replica tables.
+            qprof = QueryProfiler("SELECT date, MONTH(date) AS month,"
```
I do not like much the indentation there

jslaunwhite-microfocus replied:
I replaced this example with a better one with a different format

oualib reviewed (changes requested):
I think:
 - many docstrings are missing.
 - an entire page to explain the performance import/export should be done using the RST format. Otherwise, it will be hard to continue this part of the project.

oualib commented:
@afard please review this PR as well.
@vikash018 please look at the unit tests.

jslaunwhite-microfocus commented:
> I think:
> 
>     * many docstrings are missing.
> 

OK, I'll add more docstrings.

>     * an entire page to explain the performance import/export should be done using the RST format. Otherwise, it will be hard to continue this part of the project.

Sure, I'd be happy to add it. Where should I put the page? Should it be a comment for the import/export?

Is there some of this we can put in a follow-up task?

oualib commented:
@jslaunwhite-microfocus please look at your emailbox.

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+
+    def test_profile_export(self, amazon_vd, schema_loader, tmp_path):
+        logging.info(f"Amazon vd is {amazon_vd}")
+        request = f"""
```
I see this SQL is used in multiple test functions. It would be better to define it globally or in separate function or fixture.

jslaunwhite-microfocus replied:
The query can change if `amazon_vd` changes. I split the repeated parts into a template.

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+    Contains tests related to creating bundles of parquet data
+    """
+
+    def test_profile_export(self, amazon_vd, schema_loader, tmp_path):
```
doc string is missing in all test functions.

jslaunwhite-microfocus replied:
added docstring.

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+        assert len(set_diff) == 0
+
+    def test_profile_full_lifecycle(self, amazon_vd, schema_loader, tmp_path):
+        request = f"""
```
same

jslaunwhite-microfocus replied:
added a simple docstring

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+        return existing_tables
+
+    def test_high_level_import_export(self, amazon_vd, schema_loader, tmp_path):
+        request = f"""
```
same

jslaunwhite-microfocus replied:
added a simple docstring

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+
+        loaded_tables = self._get_set_of_tables_in_schema(schema_loader, key)
+
+        expected_tables = [
```
can we define this outside class as this is getting used multiple times?

jslaunwhite-microfocus replied:
removed the entire test where this was repeated... the test was a duplicate

vikash018 commented on verticapy/tests_new/performance/vertica/collection/test_export_simple.py:
```diff
@@ -0,0 +1,194 @@
+
+        # Also check that the tables were created ok
+        loaded_tables = self._get_set_of_tables_in_schema(schema_loader, key)
+        expected_tables = [
```
can we define this outside class as this is getting used multiple times?

jslaunwhite-microfocus replied:
This is only defined and used once now

oualib commented:
I think we have no choice except merging this PR.
